### ISSUE TITLE
File removed outside of Atom does not close tab in Atom

### ISSUE THREAD

**Issue by 50Wliu** [atom/atom#14765]
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * Reproduced the problem in Safe Mode: http://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode
    * Followed all applicable steps in the debugging guide: http://flight-manual.atom.io/hacking-atom/sections/debugging/
    * Checked the FAQs on the message board for common solutions: https://discuss.atom.io/c/faq
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=✓&q=is%3Aissue+user%3Aatom
    * Checked that there is not already an Atom package that provides the described functionality: https://atom.io/packages

### Description

Let's say I have a git repo with a single file, `a.txt`, on `master`.  The `add-b-file` branch adds a new `b.txt` file to the repo.  When on `add-b-file` with both `a.txt` and `b.txt` open in Atom, switching branches to `master` does not close `b.txt` even if there are no unsaved changes.  I have "Close Deleted File Tabs" checked.

### Steps to Reproduce

1. Create a git repo and commit a file
2. Switch to a new branch and add a different file
3. Open both files in Atom
4. Switch back to the original branch

**Expected behavior:** Only `a.txt` to exist in Atom

**Actual behavior:** Both files are still open, with `b.txt` marked as having unsaved changes.

**Reproduces how often:** 100%

### Versions
```
Atom    : 1.19.0-dev-c6bf86c9b
Electron: 1.6.9
Chrome  : 56.0.2924.87
Node    : 7.4.0
```

### Additional Information

This looks like a regression from #693.  The steps to reproduce I gave are what I believe to be a simplified version of what I'm experiencing.  If they don't actually work, the actual scenario I'm facing has to do with switching between the `wl-split-grammars` and `master` branch on atom/language-php.

Tentatively /cc'ing @iolsen because of atom/tabs#306.

**Comment by 50Wliu** [comment]
In addition, when switching back to the branch that has the additional file, the file is still marked as having unsaved changes.  Makes you double-guess about what saving will actually do...

**Comment by iolsen** [comment]
/cc @damieng @smashwilson 

This is among the categories of issues I'm hoping we resolve more robustly with the new file change API.

**Comment by backspaces** [comment]
This just caused an over-write regression in my repo. I'm not pleased, and now am concerned about keeping too many tabs open. Sigh.

Is there any workaround for this?

**Comment by ungb** [comment]
I'm able to repro with the given repro steps.  To simplify steps:

1. Run the following in your terminal
```
mkdir 14765test
cd 14765test
git init
echo "a" > a.txt
git add a.txt
git commit -m "Add a.txt"
git checkout -b b
echo "b" > b.txt
git add b.txt
git commit -m "Add b.txt"
atom .
```

2. Open both a.txt and b.txt
3. go back to terminal and run the following `git checkout master`

Expected: tab with file `b.txt` should be removed and only `a.txt` should exist

Actual: both exist, but `b.txt` is dirty.

**Comment by ungb** [comment]
I was able to verify on 1.18.0 this seems to work correctly when you have `Closed Deleted Files Tab` checked, but doesn't work off master or 1.19

**Comment by Bjvanminnen** [comment]
I can also confirm that this had been working properly for me in 1.18.0, and no longer does in 1.19. Are there any recommended steps for rolling back to an older version of atom?

**Comment by 50Wliu** [comment]
@Bjvanminnen download the previous version of Atom and install it.  Then, disable auto updates (Help -> About on Windows, Atom -> About on macOS).

**Comment by kuychaco** [comment]
Fixed in https://github.com/atom/text-buffer/pull/244

Except for the following, which is captured in a separate issue
> In addition, when switching back to the branch that has the additional file, the file is still marked as having unsaved changes. Makes you double-guess about what saving will actually do...

**Comment by joallard** [comment]
Would be fixed by d2a266e36 (unreleased)
